### PR TITLE
Generated client: embed generated methods into VMIM interface

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Before this PR:
We have our custom `virtualMachineInstanceMigrationInterface` interface with the common `Get`, `Delete`, etc...
From https://github.com/kubevirt/kubevirt/pull/11297 the common methods implementation works as a wrapper with different signatures for the generated ones, redirecting the requests to them.

After this PR:
Convert using the generated client signatures instead of the custom one. The generated `virtualMachineInstanceMigrationInterface`  is embedded into our custom one, also providing new `DeleteCollection` and `Watch` methods.
The changes inherited are:
- `Patch` and `PatchStatus` now use metav1.PatchOptions{} instead of *metav1.PatchOptions{}.
- `List` now uses metav1.ListOptions{} instead of *metav1.ListOptions{}.
- `Get` now uses metav1.GetOptions{} instead of *metav1.GetOptions{}.
- `Delete` now uses metav1.DeleteOptions{} instead of *metav1.DeleteOptions{}.
- `Create` now receives metav1.CreateOptions{} too.
- `Update` and `UpdateStatus` now receive metav1.UpdateOptions{} too.
- Added `DeleteCollection`, `Watch` methods.
- All the methods above take a  `Context` as first paramether.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
For an easier review process the commits was splitted in the following manner:
1. Align our custom common methods' signature to the generated one in the following order:
     - `Create` - introduced `CreateOptions` and introduced a `Context`
     - `Get` - used `GetOptions` instead of the pointer and introduced a `Context`
     - `List` - used `ListOptions` instead of the pointer and introduced a `Context`
     - `Update` - introduced `UpdateOptions` and introduced a `Context`
     - `Delete` - used `DeleteOptions` instead of the pointer and introduced a `Context`
     - `Patch` and `PatchStatus` - used `PatchOptions` instead of the pointer and introduced a `Context`
     - `UpdateStatus` - introduced `UpdateOptions` and introduced a `Context`
2. Embed the generated `VirtualMachineInstanceMigrationInterface` into the custom one.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] ~Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test~
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
/hold
let's wait for https://github.com/kubevirt/kubevirt/pull/11523 to reduce conflicts.
/test all

**Comment by fossedihelm** [comment]
/unhold
https://github.com/kubevirt/kubevirt/pull/11523 was merged

File: pkg/util/status/status_test.go [hunk 1]
```diff
@@ -266,7 +266,7 @@ var _ = Describe("Status", func() {
 			By("checking the VirtualMachineInstanceMigration resource")
 			migrationUpdater := NewMigrationStatusUpdater(virtClient)
 			migration := &v1.VirtualMachineInstanceMigration{Status: v1.VirtualMachineInstanceMigrationStatus{Phase: v1.MigrationPhaseUnset}}
-			migrationInterface.EXPECT().UpdateStatus(migration).Return(migration, nil).Times(1)
+			migrationInterface.EXPECT().UpdateStatus(context.Background(), migration, v12.UpdateOptions{}).Return(migration, nil).Times(1)
```
Comment (xpivarc) [inline pkg/util/status/status_test.go:269]: nit: There is `metav1`

**Review by xpivarc — COMMENTED** [review]
/approve

File: pkg/util/status/status_test.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/util/status/status_test.go]: nice catch! I removed the duplicate import in te `UpdateStatus` commit

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @fossedihelm.
Please see the inline comments

File: staging/src/kubevirt.io/client-go/kubecli/generated_mock_kubevirt.go [hunk 2]
```diff
@@ -1844,15 +1844,15 @@ func (_m *MockVirtualMachineInstanceMigrationInterface) EXPECT() *_MockVirtualMa
 	return _m.recorder
 }
 
-func (_m *MockVirtualMachineInstanceMigrationInterface) Get(name string, options *v12.GetOptions) (*v120.VirtualMachineInstanceMigration, error) {
-	ret := _m.ctrl.Call(_m, "Get", name, options)
+func (_m *MockVirtualMachineInstanceMigrationInterface) Get(ctx context.Context, name string, opts v12.GetOptions) (*v120.VirtualMachineInstanceMigration, error) {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/generated_mock_kubevirt.go]: this is generated and we cannot control the name

**Comment by fossedihelm** [comment]
@orelmisan Thank you! Addressed your comments. PTAL

**Review by orelmisan — APPROVED** [review]
Thank you for the changes

**Comment by fossedihelm** [comment]
@orelmisan Rebased because there were some conflicts in `pkg/virt-controller/watch/drain/evacuation/evacuation_test.go` PTAL

File: staging/src/kubevirt.io/client-go/kubecli/migration.go [hunk 3]
```diff
@@ -47,12 +47,8 @@ type migration struct {
...
 // Create new VirtualMachineInstanceMigration in the cluster to specified namespace
-func (o *migration) Create(newMigration *v1.VirtualMachineInstanceMigration, options *k8smetav1.CreateOptions) (*v1.VirtualMachineInstanceMigration, error) {
-	opts := k8smetav1.CreateOptions{}
-	if options != nil {
-		opts = *options
-	}
-	newMigrationResult, err := o.VirtualMachineInstanceMigrationInterface.Create(context.Background(), newMigration, opts)
+func (o *migration) Create(ctx context.Context, newMigration *v1.VirtualMachineInstanceMigration, options k8smetav1.CreateOptions) (*v1.VirtualMachineInstanceMigration, error) {
+	newMigrationResult, err := o.VirtualMachineInstanceMigrationInterface.Create(ctx, newMigration, options)
 	newMigrationResult.SetGroupVersionKind(v1.VirtualMachineInstanceMigrationGroupVersionKind)
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply staging/src/kubevirt.io/client-go/kubecli/migration.go:52]: We are relying here on an implementation detail that might change.

File: staging/src/kubevirt.io/client-go/kubecli/migration_test.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ import (
 	k8sv1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (orelmisan) [inline staging/src/kubevirt.io/client-go/kubecli/migration_test.go:33]: nit: "VMIM Update" commit: There is `k8smetav1`  one line above.

**Review by orelmisan — APPROVED** [review]
Thank you @fossedihelm

**Review by orelmisan — APPROVED** [review]

File: staging/src/kubevirt.io/client-go/kubecli/migration.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/migration.go:52]: >We are relying here on an implementation detail that might change.
@orelmisan That is true but the setting of the kind is hack as well. We will solve it properly and than this weird assumption will be dropped as well.

**Pull request merged**